lib/plug/cowboy/drainer.ex

```diff
@@ -0,0 +1,106 @@
+defmodule Plug.Cowboy.Drainer do
+  @moduledoc """
+  GenServer to drain cowboy connections at shutdown.
+
+  When starting `Plug.Cowboy` in a supervision tree, it will create a listener that receives
+  requests and creates a connection process to handle that request. During shutdown, a
+  `Plug.Cowboy` process will immediately exit, closing the listener and any open connections
+  that are still being served. In most cases, it is desireable to allow connections to
+  complete before shutting down.
+
+  This module provides a GenServer that during shutdown will close listeners and wait
+  for connections to complete. It should be placed after other supervised processes that
+  handle cowboy connections.
+
+  ## Options
+
+    * `:refs` - A list of refs to drain. `:all` is also supported and will drain all cowboy
+      listeners, including those started by means other than `Plug.Cowboy`.
+
+    * `:id` - The ID for the process.
+      Defaults to `Plug.Cowboy.Drainer`.
+
+    * `:shutdown` - How long to wait for connections to drain.
+      Defaults to 5000ms.
+
+    * `:drain_check_interval` - How frequently to check if a listener's
+      connections have been drained.
+      Defaults to 1000ms.
+
+  ## Examples
+
+      defmodule Hello.Application do
+        use Application
+        def start(_type, _args) do
+          ...
+
+          children = [
+            {Plug.Cowboy, scheme: :http, plug: MyApp, options: [port: 4040]},
+            {Plug.Cowboy, scheme: :https, plug: MyApp, options: [port: 4041]},
+            {Plug.Cowboy.Drainer, refs: [MyApp.HTTP, MyApp.HTTPS]}
+          ]
+
+          opts = [strategy: :one_for_one, name: Hello.Supervisor]
+          Supervisor.start_link(children, opts)
+        end
+      end
+  """
+  use GenServer
+
+  @doc false
+  @spec child_spec(opts :: Keyword.t()) :: Supervisor.child_spec()
```

Comment:
Wasn't sure what to write here that would be different than the module doc.

Reply:
Agreed.